python/ray/_private/utils.py

```diff
@@ -1698,3 +1698,39 @@ def _get_pyarrow_version() -> Optional[str]:
             if hasattr(pyarrow, "__version__"):
                 _PYARROW_VERSION = pyarrow.__version__
     return _PYARROW_VERSION
+
+
+class DeferSigint:
+    """Context manager that defers SIGINT signals until the the context is left."""
+
+    # This is used by Ray's task cancellation to defer cancellation interrupts during
+    # problematic areas, e.g. task argument deserialization.
+    def __init__(self):
+        # Whether the task has been cancelled while in the context.
+        self.task_cancelled = False
+        # The original SIGINT handler.
+        self.orig_sigint_handler = None
+
+    def _set_task_cancelled(self, signum, frame):
+        """SIGINT handler that defers the signal."""
+        self.task_cancelled = True
+
+    def __enter__(self):
+        # Save original SIGINT handler for later restoration.
+        self.orig_sigint_handler = signal.getsignal(signal.SIGINT)
```

Comment:
What happens if another sigint handler is set while the context manager is held?

Reply:
Thanks for writing this up, a lot of this is new to me :)

(2) seems better short term (or even a backlog item if we don't anticipate issues?)! But I thought this code is designed to protect us from interrupting non-reentrant imports. Therefore, we're running arbitrary user code during these imports, which could set up SIGINT handlers and cause issues.

Is this true, or where am I wrong? What imports are expected to run while this contextmanager is held?

Reply:
No you're correct, arbitrary user code is run during task argument deserialization, so the user could register a SIGINT signal handler; the current semantics are that the user's signal handler would take over until task deserialization is finished and the context is exited, at which point the original SIGINT signal handler is reinstated. I don't think that these semantics are the end of the world, since we're basically respecting the user's signal handler registration until task argument deserialization is done, which I think is reasonable.

So I don't anticipate issues, but I agree that (2) is a bit cleaner than either the status quo or (1). I'll take a stab at that!

Reply:
should hijack this in ray core, or only in the arrow serialization plugin?